### PR TITLE
antonio_gamez

### PR DESCRIPTION
**Pull request by antgamdia**
First attempt with exercises 01 and 02. Using a downgraded mariadb version.

### CODE REVIEW

File: antonio_gamez/session-01/exercise-01/README.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+* Use a canary deployment for WP. Consider the version 4.9.7 as the stable WP version and use 4.9.8 in the canary deployment.
+
+* Wordpress should be publicly available while MariaDB should only be accessible internally (you can consider your cluster supports LoadBalancer service type).
+
+  
+## Steps to set up this scenario
+
+1) From this directory, run `kubectl create -f ./resources` or exec the `./commands.sh` script.
+2) Open in a browser:  http://18.213.63.179:30080 or  https://18.213.63.179:30443.
```
Comment (juan131) [inline antonio_gamez/session-01/exercise-01/README.md]: Is that your Sandbox IP? Please do not hardcode it on a "generic" solution

File: antonio_gamez/session-01/exercise-02/README.md [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+3) Enjoy your new k8s-flavored Wordpress :)
+
+## Remarks about the proposed solution
+
+Discussion about some non-trivial aspects (not covered in exercise01) of the solution and the expected goals:
+
+> MariaDB should be configured with replication with a master-slave model. There should be 1 master and 2 slaves.
+
+Two different deployments have been created, master and slave. The main differences are the set of environment variables and the number of initial replicas.
+Due to a bug in the latest bitnami/mariadb image related to the proper configuration of the env vars, the version has been downgraded to `10.1.21-r2`.
```
Comment (juan131) [inline antonio_gamez/session-01/exercise-02/README.md]: Not necessary if you configure the slave pods correctly

File: antonio_gamez/session-01/exercise-02/README.md [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+  strategy:
+    type: RollingUpdate
+    rollingUpdate:
+      maxUnavailable: 40%
+      maxSurge: 50%
+```
+
+> Scale slaves to 5 replicas. You can document how to do it or include the steps in the *commands.bash* file.
+
+It is achieved simply typing `kubectl scale -l "app=wordpress,tier=database,mode=slave" --replicas=5`. A best approach to do scaling up/down is to define an Horizontal Pod Autoscaler from some default (or even custom) metrics. A simple autoscaler could be `kubectl autoscale -l "app=wordpress,tier=database,mode=slave" --min=2 --max=5 --cpu-percent=80`
```
Comment (juan131) [inline antonio_gamez/session-01/exercise-02/README.md]: Nice explanation!

File: antonio_gamez/session-01/exercise-02/README.md [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+  
+This scenario is *absolutely* not ready for production purposes. Specifically, there remain two major issues to be addressed:
+
+1) **Expose the service in a common 80 port**. Two alternatives here: 
+    1.1) Use **LoadBalancer** service: in order to use a LoadBalancer in AWS, we should move to an EKS cluster for that, or install the AWS addon that allows you to create ELBs on demand.
+    1.1) Use an **Ingress** controller: we have to install the Ingress addon into the cluster (e.g. nginx or HAProxy) and create the rules to forward the traffic
+
+2) **Configure SSL/TLS to secure the http traffic**. A straightforward way to do that is to generate a Let'sEncrypt free certificate and configure it in either the WordPress or the ingress service.
+
+3) **Add persistence** by configuring NFS/EBS/... storage and creating PVCs.
```
Comment (juan131) [inline antonio_gamez/session-01/exercise-02/README.md]: It'll be addressed on other sessions.

File: antonio_gamez/session-01/exercise-02/README.md [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+4) **Consider using StatefulSet** when adding persistence to nodes. This object deals with a distributed application that requires each node to have a persistent state and the ability to configure an arbitrary number of nodes through a configuration
```
Comment (juan131) [inline antonio_gamez/session-01/exercise-02/README.md]: It'll be addressed on other sessions.

File: antonio_gamez/session-01/exercise-02/README.md [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+    rollingUpdate:
+      maxUnavailable: 40%
+      maxSurge: 50%
+```
+
+> Scale slaves to 5 replicas. You can document how to do it or include the steps in the *commands.bash* file.
+
+It is achieved simply typing `kubectl scale -l "app=wordpress,tier=database,mode=slave" --replicas=5`. A best approach to do scaling up/down is to define an Horizontal Pod Autoscaler from some default (or even custom) metrics. A simple autoscaler could be `kubectl autoscale -l "app=wordpress,tier=database,mode=slave" --min=2 --max=5 --cpu-percent=80`
+
+> Document the steps to follow to install HyperDB WP plugin and configure it to balance SQL request between Master&Slaves services.
```
Comment (juan131) [inline antonio_gamez/session-01/exercise-02/README.md]: Nice!!

File: antonio_gamez/session-01/exercise-02/resources/mariadb-master/mariadb-cm.yaml [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: mariadb-cm
+  namespace: exercise-02
+  labels:
+    app: wordpress
+    tier: database
+    stage: production
```
Comment (juan131) [inline antonio_gamez/session-01/exercise-02/resources/mariadb-master/mariadb-cm.yaml]: No canary scenario, please remove it

File: antonio_gamez/session-01/exercise-02/resources/mariadb-master/mariadb-master-deploy.yaml [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+kind: Deployment
+apiVersion: extensions/v1beta1
+metadata:
+  name: mariadb-master-deploy
+  namespace: exercise-02
+  labels:
+    app: wordpress
+    tier: database
+    mode: master
+    stage: production
```
Comment (juan131) [inline antonio_gamez/session-01/exercise-02/resources/mariadb-master/mariadb-master-deploy.yaml]: No canary scenario, please remove it

File: antonio_gamez/session-01/exercise-02/resources/mariadb-master/mariadb-master-svc.yaml [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+apiVersion: v1
+kind: Service
+metadata:
+  name: mariadb-master
+  namespace: exercise-02
+  labels:
+    app: wordpress
+    tier: database
+    mode: master
+    stage: production
```
Comment (juan131) [inline antonio_gamez/session-01/exercise-02/resources/mariadb-master/mariadb-master-svc.yaml]: No canary scenario, please remove it

File: antonio_gamez/session-01/exercise-02/resources/mariadb-slave/mariadb-slave-deploy.yaml [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+kind: Deployment
+apiVersion: extensions/v1beta1
+metadata:
+  name: mariadb-slave-deploy
+  namespace: exercise-02
+  labels:
+    app: wordpress
+    tier: database
+    mode: slave
+    stage: production
```
Comment (juan131) [inline antonio_gamez/session-01/exercise-02/resources/mariadb-slave/mariadb-slave-deploy.yaml]: No canary scenario, please remove it

File: antonio_gamez/session-01/exercise-02/resources/mariadb-slave/mariadb-slave-svc.yaml [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+apiVersion: v1
+kind: Service
+metadata:
+  name: mariadb-slave
+  namespace: exercise-02
+  labels:
+    app: wordpress
+    tier: database
+    mode: slave
+    stage: production
```
Comment (juan131) [inline antonio_gamez/session-01/exercise-02/resources/mariadb-slave/mariadb-slave-svc.yaml]: No canary scenario, please remove it

File: antonio_gamez/session-01/exercise-02/resources/wordpress/wordpress-cm.yaml [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: wordpress-cm
+  namespace: exercise-02
+  labels:
+    app: wordpress
+    tier: frontend
+    stage: production
```
Comment (juan131) [inline antonio_gamez/session-01/exercise-02/resources/wordpress/wordpress-cm.yaml]: No canary scenario, please remove it

File: antonio_gamez/session-01/exercise-02/resources/wordpress/wordpress-deploy.yaml [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+############################ # ############################
+---
+kind: Deployment
+apiVersion: extensions/v1beta1
+metadata:
+  name: wordpress-deploy
+  namespace: exercise-02
+  labels:
+    app: wordpress
+    stage: production
```
Comment (juan131) [inline antonio_gamez/session-01/exercise-02/resources/wordpress/wordpress-deploy.yaml]: No canary scenario, please remove it

File: antonio_gamez/session-01/exercise-02/resources/wordpress/wordpress-pvc.yaml [hunk 14]
```diff
@@ -0,0 +1,26 @@
+# ############################ # ############################
```
Comment (juan131) [inline antonio_gamez/session-01/exercise-02/resources/wordpress/wordpress-pvc.yaml]: Nice. However, it's not requested for this exercise

File: antonio_gamez/session-01/exercise-02/resources/wordpress/wordpress-deploy.yaml [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+      - name: wordpress
+        image: bitnami/wordpress:4.9.8
+        ports:
+        - name: http-port
+          containerPort: 80
+          protocol: TCP
+        - name: https-port
+          containerPort: 443
+          protocol: TCP
+        resources:
```
Comment (juan131) [inline antonio_gamez/session-01/exercise-02/resources/wordpress/wordpress-deploy.yaml]: Nicee!

File: antonio_gamez/session-01/exercise-02/resources/wordpress/wordpress-deploy.yaml [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+            port: 80
+            scheme: HTTP
+          initialDelaySeconds: 30
+          timeoutSeconds: 5
+          periodSeconds: 10
+          successThreshold: 1
+          failureThreshold: 6
+      restartPolicy: Always
+      terminationGracePeriodSeconds: 30
+  strategy:
```
Comment (juan131) [inline antonio_gamez/session-01/exercise-02/resources/wordpress/wordpress-deploy.yaml]: Really cool!

File: antonio_gamez/session-01/exercise-02/resources/mariadb-slave/mariadb-slave-deploy.yaml [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+            - "mysqladmin ping -u root -p${MYSQL_ROOT_PASSWORD}"
+          initialDelaySeconds: 120
+          timeoutSeconds: 1
+          periodSeconds: 10
+          successThreshold: 1
+          failureThreshold: 6
+        imagePullPolicy: Always
+      restartPolicy: Always
+      terminationGracePeriodSeconds: 30
+  strategy:
```
Comment (juan131) [inline antonio_gamez/session-01/exercise-02/resources/mariadb-slave/mariadb-slave-deploy.yaml]: Really cool!

File: antonio_gamez/session-01/exercise-02/resources/mariadb-master/mariadb-master-deploy.yaml [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+            - "mysqladmin ping -u root -p${MYSQL_ROOT_PASSWORD}"
+          initialDelaySeconds: 120
+          timeoutSeconds: 1
+          periodSeconds: 10
+          successThreshold: 1
+          failureThreshold: 6
+        imagePullPolicy: Always
+      restartPolicy: Always
+      terminationGracePeriodSeconds: 30
+  strategy:
```
Comment (juan131) [inline antonio_gamez/session-01/exercise-02/resources/mariadb-master/mariadb-master-deploy.yaml]: Cool!

File: antonio_gamez/session-01/exercise-02/resources/mariadb-master/mariadb-master-deploy.yaml [hunk 19]
```diff
@@ -0,0 +1,118 @@
...
+        stage: production
+    spec:
+      containers:
+      - name: mariadb
+        image: bitnami/mariadb:10.1.21-r2 # see https://github.com/bitnami/bitnami-docker-mariadb
+        ports:
+        - name: db-port
+          containerPort: 3306
+          protocol: TCP
+        resources:
```
Comment (juan131) [inline antonio_gamez/session-01/exercise-02/resources/mariadb-master/mariadb-master-deploy.yaml]: 👍

File: antonio_gamez/session-01/exercise-02/resources/mariadb-master/mariadb-master-deploy.yaml [hunk 20]
```diff
@@ -0,0 +1,118 @@
...
+    metadata:
+      labels:
+        app: wordpress
+        tier: database
+        mode: master
+        stage: production
+    spec:
+      containers:
+      - name: mariadb
+        image: bitnami/mariadb:10.1.21-r2 # see https://github.com/bitnami/bitnami-docker-mariadb
```
Comment (juan131) [inline antonio_gamez/session-01/exercise-02/resources/mariadb-master/mariadb-master-deploy.yaml]: Use latest!

File: antonio_gamez/session-01/exercise-02/resources/mariadb-slave/mariadb-slave-deploy.yaml [hunk 21]
```diff
@@ -0,0 +1,160 @@
...
+    metadata:
+      labels:
+        app: wordpress
+        tier: database
+        mode: slave
+        stage: production
+    spec:
+      containers:
+      - name: mariadb
+        image: bitnami/mariadb:10.1.21-r2 # see https://github.com/bitnami/bitnami-docker-mariadb
```
Comment (juan131) [inline antonio_gamez/session-01/exercise-02/resources/mariadb-slave/mariadb-slave-deploy.yaml]: Use latest!

File: antonio_gamez/session-01/exercise-02/resources/mariadb-slave/mariadb-slave-deploy.yaml [hunk 22]
```diff
@@ -0,0 +1,160 @@
...
+        ports:
+        - name: db-port
+          containerPort: 3306
+          protocol: TCP
+        resources:
```
Comment (juan131) [inline antonio_gamez/session-01/exercise-02/resources/mariadb-slave/mariadb-slave-deploy.yaml]: 👍

File: antonio_gamez/session-01/exercise-02/resources/mariadb-slave/mariadb-slave-deploy.yaml [hunk 23]
```diff
@@ -0,0 +1,160 @@
...
+            requests:
+              memory: "256m"
+              cpu: "250m"
+            limits:
+              memory: "512Mi"
+              cpu: "500m"
+        env:
+          - name: MARIADB_USER
```
Comment (juan131) [inline antonio_gamez/session-01/exercise-02/resources/mariadb-slave/mariadb-slave-deploy.yaml]: Not necessary on slaves

File: antonio_gamez/session-01/exercise-02/resources/mariadb-slave/mariadb-slave-deploy.yaml [hunk 24]
```diff
@@ -0,0 +1,160 @@
...
+            valueFrom:
+              configMapKeyRef:
+                name: mariadb-cm
+                key: MARIADB_USER
+          - name: MARIADB_PASSWORD
```
Comment (juan131) [inline antonio_gamez/session-01/exercise-02/resources/mariadb-slave/mariadb-slave-deploy.yaml]: Not necessary on slaves

File: antonio_gamez/session-01/exercise-02/resources/mariadb-slave/mariadb-slave-deploy.yaml [hunk 25]
```diff
@@ -0,0 +1,160 @@
...
+            valueFrom:
+              secretKeyRef:
+                name: mariadb-secret # from secret
+                key: MARIADB_PASSWORD
+          - name: MARIADB_DATABASE
```
Comment (juan131) [inline antonio_gamez/session-01/exercise-02/resources/mariadb-slave/mariadb-slave-deploy.yaml]: Not necessary on slaves

File: antonio_gamez/session-01/exercise-02/resources/mariadb-slave/mariadb-slave-deploy.yaml [hunk 26]
```diff
@@ -0,0 +1,160 @@
...
+            valueFrom:
+              configMapKeyRef:
+                name: mariadb-cm
+                key: MARIADB_REPLICATION_USER
+          - name: MARIADB_REPLICATION_PASSWORD
+            valueFrom:
+              secretKeyRef:
+                name: mariadb-secret # from secret
+                key: MARIADB_REPLICATION_PASSWORD
+          - name: MARIADB_MASTER_USER # old env var
```
Comment (juan131) [inline antonio_gamez/session-01/exercise-02/resources/mariadb-slave/mariadb-slave-deploy.yaml]: Remove this env. var

File: antonio_gamez/session-01/exercise-02/resources/mariadb-slave/mariadb-slave-deploy.yaml [hunk 27]
```diff
@@ -0,0 +1,160 @@
...
+              secretKeyRef:
+                name: mariadb-secret # from secret
+                key: MARIADB_ROOT_PASSWORD
+          - name: MARIADB_MASTER_ROOT_PASSWORD
+            valueFrom:
+              secretKeyRef:
+                name: mariadb-secret # from secret
+                key: MARIADB_ROOT_PASSWORD # matches general root
+          # problematic env var              
+          - name: MARIADB_ROOT_USER
```
Comment (juan131) [inline antonio_gamez/session-01/exercise-02/resources/mariadb-slave/mariadb-slave-deploy.yaml]: Not necessary on slaves

File: antonio_gamez/session-01/exercise-02/resources/mariadb-slave/mariadb-slave-deploy.yaml [hunk 28]
```diff
@@ -0,0 +1,160 @@
...
+            valueFrom:
+              configMapKeyRef:
+                name: mariadb-cm
+                key: MARIADB_MASTER_ROOT_USER
+          # problematic env var
+          - name: MARIADB_ROOT_PASSWORD
```
Comment (juan131) [inline antonio_gamez/session-01/exercise-02/resources/mariadb-slave/mariadb-slave-deploy.yaml]: Not necessary on slaves

File: antonio_gamez/session-01/exercise-02/resources/mariadb-slave/mariadb-slave-deploy.yaml [hunk 29]
```diff
@@ -0,0 +1,160 @@
...
+            valueFrom:
+              secretKeyRef:
+                name: mariadb-secret # from secret
+                key: MARIADB_ROOT_PASSWORD # matches general root
+        livenessProbe:
+          exec:
+            command:
+            - sh
+            - -c
+            - "mysqladmin ping -u root -p${MYSQL_ROOT_PASSWORD}"
```
Comment (juan131) [inline antonio_gamez/session-01/exercise-02/resources/mariadb-slave/mariadb-slave-deploy.yaml]: Use `MARIADB_MASTER_ROOT_PASSWORD`

File: antonio_gamez/session-01/exercise-02/resources/mariadb-slave/mariadb-slave-deploy.yaml [hunk 30]
```diff
@@ -0,0 +1,160 @@
...
+          timeoutSeconds: 5
+          periodSeconds: 10
+          successThreshold: 1
+          failureThreshold: 6
+        readinessProbe:
+          exec:
+            command:
+            - sh
+            - -c
+            - "mysqladmin ping -u root -p${MYSQL_ROOT_PASSWORD}"
```
Comment (juan131) [inline antonio_gamez/session-01/exercise-02/resources/mariadb-slave/mariadb-slave-deploy.yaml]: Use `MARIADB_MASTER_ROOT_PASSWORD`

File: antonio_gamez/session-01/exercise-01/resources/wordpress/wordpress-cm.yaml [hunk 31]
```diff
@@ -0,0 +1,31 @@
...
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: wordpress-cm
+  namespace: exercise-01
+  labels:
+    app: wordpress
+    tier: frontend
+    stage: production
+data:
```
[the remark this answers is not in the document]
Comment (antgamdia) [reply antonio_gamez/session-01/exercise-01/resources/wordpress/wordpress-cm.yaml]: Totally, username and db user should be moved to a secret. For these dummy scenarios, I considered as 'secret' just the passwords, but there exist a lot of attack vectors based on other sensitive data.

Regarding the placing of the variables, I followed a simple rule: if it's something I'd introduce in WP, I've created it inside its corresponding secret/cm, regardless of the semantic of the parameter. I believe it's a simplier approach when having a large ecosystem of services, but final decision belongs to the team's majority opinion.

File: antonio_gamez/session-01/exercise-01/README.md (hunk 1 above)
Comment (antgamdia) [reply antonio_gamez/session-01/exercise-01/README.md]: I misunderstood the target of the instructions. Fixing it right now. Thanks!

File: antonio_gamez/session-01/exercise-01/README.md [hunk 32]
```diff
@@ -0,0 +1,52 @@
...
+
+## Technical debt
+  
+This scenario is *absolutely* not ready for production purposes. Specifically, there remain two major issues to be addressed:
+
+1) **Expose the service in a common 80 port**. Two alternatives here: 
+    1.1) Use **LoadBalancer** service: in order to use a LoadBalancer in AWS, we should move to an EKS cluster for that, or install the AWS addon that allows you to create ELBs on demand.
+    1.1) Use an **Ingress** controller: we have to install the Ingress addon into the cluster (e.g. nginx or HAProxy) and create the rules to forward the traffic
+
+2) **Configure SSL/TLS to secure the http traffic**. A straightforward way to do that is to generate a Let'sEncrypt free certificate and configure it in either the wordpress or the ingress service.
```
[the remark this answers is not in the document]
Comment (antgamdia) [reply antonio_gamez/session-01/exercise-01/README.md]: Mmm, but for managing certificates directly in the EC2 we need to use the AWS Certificate Manager (disabled for our accounts), isn't right?

**Comment by antgamdia** [comment]
Btw, a LoadBalancer (manually) is configured and working at: http://k8s.governify.io (available only during the instance working hours 💰 )

**Comment by juan131** [comment]
Please remove the **LoadBalancer**, it's not requested for the exercise. We just wanted you to play with the svc configuration.

**Comment by antgamdia** [comment]
> Please remove the **LoadBalancer**, it's not requested for the exercise. We just wanted you to play with the svc configuration.

Done!

**Comment by juan131** [comment]
Thanks 👍

File: antonio_gamez/session-02/exercise-01/README.md [hunk 33]
```diff
@@ -0,0 +1,91 @@
...
+* Add a TLS certificate
+* Force HTTPS redirection
+
+
+
+
+
+
+
+COSAS
```
Comment (juan131) [inline antonio_gamez/session-02/exercise-01/README.md]: ????

File: antonio_gamez/session-02/exercise-01/README.md [hunk 34]
```diff
@@ -0,0 +1,91 @@
...
+
+1. .spec.updateStrategy antes no existia y el comportaiiento era otro (ondelete)
```
Comment (juan131) [inline antonio_gamez/session-02/exercise-01/README.md]: ?????

**Comment by antgamdia** [comment]
I've just reverted all the session02-related commits. Sorry!

**Pull request merged**